flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AsyncIntervalJoinOperator.java

```diff
@@ -0,0 +1,426 @@
+                                final long timestamp = bucket.getKey();
+
+                                if (timestamp < ourTimestamp + relativeLowerBound
+                                        || timestamp > ourTimestamp + relativeUpperBound) {
+                                    continue;
+                                }
+
+                                for (IntervalJoinOperator.BufferEntry<OTHER> entry :
+                                        bucket.getValue()) {
+                                    if (isLeft) {
+                                        collect(
+                                                (T1) ourValue,
+                                                (T2) entry.getElement(),
+                                                ourTimestamp,
+                                                timestamp);
+                                    } else {
+                                        collect(
+                                                (T1) entry.getElement(),
+                                                (T2) ourValue,
+                                                timestamp,
+                                                ourTimestamp);
+                                    }
+                                }
+                            }
+
+                            long cleanupTime =
+                                    (relativeUpperBound > 0L)
+                                            ? ourTimestamp + relativeUpperBound
+                                            : ourTimestamp;
+                            if (isLeft) {
+                                internalTimerService.registerEventTimeTimer(
+                                        CLEANUP_NAMESPACE_LEFT, cleanupTime);
+                            } else {
+                                internalTimerService.registerEventTimeTimer(
+                                        CLEANUP_NAMESPACE_RIGHT, cleanupTime);
+                            }
+                        });
+    }
+
+    private boolean isLate(long timestamp) {
+        long currentWatermark = internalTimerService.currentWatermark();
+        return timestamp < currentWatermark;
+    }
+
+    /** Write skipped late arriving element to SideOutput. */
+    protected <T> void sideOutput(T value, long timestamp, boolean isLeft) {
+        if (isLeft) {
+            if (leftLateDataOutputTag != null) {
+                output.collect(leftLateDataOutputTag, new StreamRecord<>((T1) value, timestamp));
+            }
+        } else {
+            if (rightLateDataOutputTag != null) {
+                output.collect(rightLateDataOutputTag, new StreamRecord<>((T2) value, timestamp));
+            }
+        }
+    }
+
+    private void collect(T1 left, T2 right, long leftTime, long rightTime) throws Exception {
+        resultTimestamp.set(Math.max(leftTime, rightTime));
+        leftTimestamp.set(leftTime);
+        rightTimestamp.set(rightTime);
+
```

Comment:
Should context be updated here ? 
```
 context.updateTimestamps(leftTimestamp, rightTimestamp, resultTimestamp);
```

Reply:
No need. The internal variables have been updated here `result/left/rightTimestamp`, which will be used by the context